
<html>
<head>
<title>September, 2005: C++ Exceptions &amp; the  Linux Kernel</title>
</head>

<body BGCOLOR="#ffffff" LINK="#0000ff" VLINK="#330066" ALINK="#ff0000" TEXT="#000000">
<!--Copyright &#169; Dr. Dobb's Journal-->

<h1>C++ Exceptions &amp; the  Linux Kernel</h1>
<p><i>Dr. Dobb's Journal</i> September, 2005</p>
<h2>Leveraging the  power of C++</h2>


<h3>By Halld&oacute;r &Iacute;sak Gylfason and G&iacute;sli Hj&aacute;lmtysson</h3>


<I>Halld&oacute;r was a research assistant at the Network Systems and Services lab at Reykjavik University and chief architect at Calidris. He is currently a Ph.D. student at the University of California at Berkeley. G&iacute;sli is a professor of computer science at Reykjavik University. He previously was a member of the technical staff at AT&amp;T Bell Labs, where he researched networking systems and services, optical networking, router architectures, modeling, and performance evaluations. They can be contacted at halldorisak@ru.is and gisli@ru.is, respectively.</I>

<hr>





<p>Exceptions are common in user-space programming. Most modern programming languages offer some form of syntactic constructs to handle exceptional events. The belief is widespread that the use of exceptions leads to more maintainable and robust systems, as error-handling code is separated from the normal flow. Multiple modern programming styles and best practices encourage the use of exceptions to handle exceptional cases. When a function detects an error, an exception is raised, which directs the program flow to the nearest dynamically enclosing handler for that type of exception. Thus, exceptions are handled differently from normal procedure exits, and exceptions are transparently passed through functions that do not handle the error.</p>

<p>In spite of their common use in user space, the use of exceptions in kernel space has been limited. In fact, some operating systems do not exploit higher level language abstractions at all. In particular, Linux is written in pure C. Whereas performance issues may negate the use of some modern languages such as Java, one of the driving factors behind the creation of C++ was for use in writing operating systems. Some constructs were specifically introduced and designed based on observed patterns&#151;and to address problems&#151;in operating system implementations. Although C++ does not enforce strong type safety or other safety properties (such as Java), C++ offers an array of high-level language abstractions valuable for the construction of operating systems, and offers type safety and compiler support far beyond that of C. In particular, the safety provided by language-level polymorphism provides significant value as polymorphic behavior is widespread throughout any operating system. </p>

<p>In our work on the Pronto software router (http://netlab.ru.is/pronto/pronto .shtml), we have used many of the advanced C++ constructs extensively, including classes and virtual functions to achieve clarity, flexibility, and extensibility. We have shown that these benefits come at no performance penalty compared to the Linux implementation (see "The Pronto Platform: A Flexible Toolkit for Programming Networks using a Commodity Operating System" by G. Hj&aacute;lmtysson. <i>Proceedings of OpenArch 2000</i>, March 2000). Our desire to employ the full range of C++ abstractions in the kernel and, in particular, to use C++ exceptions in our work on Pronto is the driver behind the work presented herein.</p>

<p>Of course, handling exceptional conditions is relatively expensive, regardless of the mechanisms employed for implementation. Substantial fraction of operating system code (for any operating system) is there to resolve exceptional conditions. Handling such conditions requires:</p>



<ul>
  <li>Detecting when an exceptional condition occurs.</li>
  <li>Determining where (by whom) such conditions should be handled.</li>
  <li>Doing the work needed to recover resources and otherwise handle the condition to return the operating system to a state from where it is safe to resume normal execution.</li>
</ul>



<p>The cost of the first and the last requirements are independent of the mechanisms employed to implement the second. The use of language-level exception handling translates into machinery providing complete and systematic approach to the second&#151;to identify where a thrown exception should be handled. </p>

<p>The performance cost of using the language-level exception machinery must be weighed against both its nonperformance benefits and the total cost of handling a given exceptional condition. Important nonperformance quality metrics include reliability, robustness, flexibility, maintainability, and speed of development. In contrast, ad hoc exception-handling patterns, common particularly in the Linux kernel, consist of convoluted traces where exceptional function abort is communicated to the caller via an exceptional return value. The performance overhead of throwing exceptions in C++ is appreciable when compared, for example, to a simple return with an integer error code. However, in many cases, executing the exception-handling code dominates the total cost of recovering from an error condition. In those cases, the substantial benefits of exceptions warrant the relatively small cost. </p>

<p>Clearly, throwing a C++ exception is more expensive than returning from a function. Therefore, there are cases where using exceptions should be avoided. The latter would typically apply when the exception handling is trivial and a simple return value is appropriate, or for exceptions that occur relatively frequently (and thus, perhaps constitute a branch rather than a true exception) or operate on such a fine time-scale that even a small overhead is burdensome. However, rather than voiding the viability of using language-level exceptions, the added cost instead determines the granularity appropriate for the use of exceptions, and/or the rarity at which the exceptional case must occur to justify the cost. </p>
<h3>Implementation of Exception Handling</h3>

<p>There are two main approaches to implementing exception handling in modern languages such as C++ (see "Exception Handling for C++," by Andrew Koenig and Bjarne Stroustrup. <i>Journal of Object Oriented Programming</i>, July/August 1990). The first approach employs dynamic registration based on the <i>setjmp</i>/<i>longjmp</i> methods. This method manages a stack of execution contexts&#151;each entry to a <i>try-block</i> pushes an execution context onto the stack through the <i>setjmp</i> method, while each exit pops an execution context off the stack through the <i>longjmp</i> method. This approach incurs some (albeit small) cost when entering <i>try</i> blocks, but has the advantage that it is portable in the sense that it is possible to generate ANSI C from a C++ program. </p>

<p>The second approach is based on a statically generated table of program counter values that map <i>try</i> blocks into program counter values for exception handlers. When an exception is thrown, this table is searched for the appropriate handler using the current program counter value as entry point. This method incurs no runtime overhead when entering a <i>try</i> block (zero instructions), at the expense of increased overhead when throwing and during exception.</p>

<p>The current GNU g++ compiler implementation, on which we have built our runtime library, implements exception handling using the table-driven approach. Former versions of the compiler used the <i>setjmp</i>/<i>longjmp</i> method, and that version can still be compiled into the implementation. The GNU implementation is contained in the (user-level) application binary interface (ABI), which accompanies the compiler as part of the standard runtime library. When using C++ exceptions, GNU g++ generates calls to the ABI; for example, the <i>throw</i> operator is transformed into a call to the ABI function <i>__cxa_allocate_exception</i> followed by <i>__cxa_throw</i>. Versions 3.x of GNU g++ implements the C++ ABI specification for the IA-64 (see "C++ ABI for Itanium: Exception Handling," http:// www.codesourcery.com/cxx-abi/abi-eh.html). The aim of the C++ ABI specification is to standardize the object layout and the interface of the object code to the runtime system. Thus, code compiled with old versions of the compiler should be compatible with newer releases, and object code from different compilers should be compatible.</p>

<p>Employing a language-level exception mechanism at kernel level has not become common in practice. Some operating systems are written at least partly in C++ but generally do not employ C++ exceptions. However, Windows NT provides a facility called "Structured Exception Handling" (SEH) (see "A Crash Course on the Depths of Win32TM Structured Exception Handling," by Matt Pietrek. <i>Microsoft System Journal</i>, January 1997), which can be used in kernel device drivers. SEH is an operating-system facility and is, therefore, independent of any compiler or language. SEH uses dynamic registrations of <i>try</i> blocks and thus does affect normal program flow. Each thread is associated with a stack of exception registrations, which contain a function pointer to a handler function. Although SEH is similar to C++ exception handling, it has different semantics. In the SEH model, exceptions are thrown explicitly with the <i>RaiseException</i> Win32 routine and&#151;in contrast to the C++ exception model&#151;exceptions in the SEH model are singular integers. However, the SEH model also covers processor-level errors, such as divide-by-zero, access violation, and stack overflow, which requires support from the OS. </p>

<p>When an exception is raised in the SEH model, the operating system calls the handler functions on the exception registration stack in sequence. Each handler function decides whether to handle the exception, to pass it through, or to resume execution at the point where the exception was raised. One drawback of SEH is that it does not call class destructors during stack unwinding; however, the SEH-specific <i>__finally</i> block can be used to clean up resources. The Microsoft C++ compiler implements C++ exceptions on top of the SEH model, and as a consequence, a <i>catch(...)</i> block catches all C++ exceptions as well as processor-level errors, such as segmentation faults. Segmentation faults in Linux are not caught when compiled with the GNU g++ compiler.</p>
<h3>Implementing the  C++ Runtime Support</h3>

<p>The first step in our work to support kernel-level exceptions was to create and include an implementation of the C++ ABI in the kernel. We started by carving out the user-level GNU ABI implementation as the basis for our implementation; however, simply porting user-level code to the kernel requires some changes. The <i>malloc</i> function used to reserve space on the heap for exceptions is replaced with the kernel-level <i>kmalloc</i> function. Furthermore, the GNU library is threadsafe and uses the pthread library for locking. However, as the pthread library is not available in kernel space, we modified the locking mechanism to use kernel-level spinlocks. </p>

<p>The semantics of the C++ <i>throw</i> operator requires the implementation to keep a stack of active exceptions to support the usage of <i>throw</i> without any arguments. In user space, this is done using thread-local storage. To achieve this effect, in kernel space, we augmented the Linux <i>task_struct</i> to include information on the active exceptions; see <a name="rl1"></a><a href="#l1">Listing One</a>.</p>

<p>When creating ELF executables, GNU g++ silently links in two object files at the front and the back (crtbegin.o and crtend.o, respectively). Furthermore, GNU g++ adds initialization code into the ELF .init section, and clean-up code into the .fini section. This is necessary to ensure that global constructors and destructors are run, and that the exception tables are registered with the ABI. To enable the usage of exceptions and global objects in the kernel, we modified the Linux makefile rule for the kernel image and kernel modules to link with those two files. Furthermore, we ensure that the initialization routines are called on module load by clever use of preprocessor macros, which modify the definition of the module initialization and finalization functions, <i>module_init</i> and <i>module_exit</i>. This is necessary because the kernel module loader in Linux pays no attention to the ELF .init section.</p>

<p>We implemented several optimizations to the standard GNU implementation of exceptions because our original measurements indicated that the cost of using exceptions was somewhat high. One optimization concerns the fact that the standard GNU implementation unwinds the stack in two phases&#151;the first phase locates a handler, and the second phase unwinds the stack to the handler. The rationale for this two-phased approach is that, in the case that no handler is found, the stack frames have not been destroyed and debuggers can inspect the state of the frame that threw the exception. However, for our use at kernel level, we don't see this cost justified. In fact, we feel that even in user space, programmers should have the option of having the compiler optimize this debugging help out of the code. </p>

<p>To enhance the performance of exceptions further, we included an optimization that improves the mechanisms used to search for the handler of an exception by caching frame state of functions, which is used to restore registers that have been saved on the stack. Our optimization caches this frame state data in a hash table, indexed by program counter. When an exception is thrown the first time through a function, or more specifically, the first time through a certain place in the function, the frame state is computed and subsequently inserted into the hash table. Subsequent throws through this place result in a successful lookup in the hash table. The importance of this optimization increases as more exceptions are used in the kernel. This is because the time needed to locate the frame descriptor entry for a function is proportional to the number of modules that use exceptions and the number of functions within those modules.</p>

<p>The cost of dynamic type checking in C++ is highly dependent on the method used to encode the runtime type information in the objects. GNU g++ follows the traditional approach and associates with each class a type information object that encodes the type of the class as a mangled string and puts a pointer to this object in the virtual table for the class. GNU g++ uses weak symbols to reduce the dynamic type checking to a pointer comparison, thus avoiding the more expensive string comparison. Each time a class containing virtual functions is used in a source file, GNU g++ generates the virtual table, type information object, and type name string as weak symbols. The user-space linker, ld, ensures that there is only one copy of this object, which renders the simple pointer comparison sufficient. However, the kernel module loader, which in the 2.6 versions of the kernel is exclusively in kernel space, does not handle these weak symbols and always relocates references to weak symbols to the weak definition within each object file. Therefore, multiple type information objects may exist for the same class, and pointer comparison becomes insufficient when doing dynamic type checking across kernel modules. To avoid this overhead, we modified the Linux kernel module loader to handle these weak symbols; the first time a weak symbol is encountered, it is added to the symbol map, and on subsequent encounters, the relocation is done relative to the first symbol. </p>
<h3>Using Kernel-Level Exceptions</h3>

<p>The C++ kernel-level runtime support for Linux provides complete runtime support for C++, including support for virtual functions, memory allocation operators, global constructors/destructors, dynamic type checking, and exceptions. The code is installed by applying a patch to the Linux kernel and enables the full use of C++ using the GNU g++ compiler. </p>

<p>Using our new C++ kernel-level runtime support, programming in C++ at kernel level becomes similar to programming in user space. The compiler compiles files with the suffix .cc as a C++ file. However, the Linux kernel distribution is written in vanilla C, so consequently, C++ source files need to include C files to interface with the Linux core. This introduces a problem not commonly encountered in user space, as some of the C++ keywords have been used as identifiers in the Linux header files. To combat this, we have provided two inclusion files&#151;begin_include.h and end_include.h&#151;with our distribution that should be used to enclose the Linux C header files, as in <a name="rl2"></a><a href="#l2">Listing Two</a>. These two files use <i>#define</i> and <i>#undef</i>, respectively, to redefine these identifiers to names accepted by the C++ compiler.</p>

<p>In the following examples, we use a class hierarchy (<a name="rl3"></a><a href="#l3">Listing Three</a>) where we group the exceptions according to sub-systems, using inheritance. The top-level exception class&#151;<i>OSException</i>&#151;consists of a message, severity, and the virtual method <i>report</i> that, by default, prints the message through <i>printk</i> if the severity is MAJOR or FATAL. The other two exceptions defined are <i>NetworkException,</i> derived from <i>OSException</i>, and <i>ProntoException,</i> derived from <i>NetworkException</i>.</p>
<h3>System Calls</h3>

<p>The most straightforward use of exceptions in kernel space is in system calls. The Pronto architecture introduces three new system calls to the Linux kernel. New types of packet processors (an abstract data type for processing in the data path of the Pronto router) can be plugged into the operating system at runtime and their behavior manipulated through type-specific system calls that are dynamically linked through virtual functions. To promote safety, it is beneficial to catch all exceptions thrown by packet processors.</p>

<p><a name="rl4"></a><a href="#l4">Listing Four</a> shows the use of exceptions to guard a system call. In this example, the <i>sys_pproc_type_call</i> is the entry point from the system call. Its only function is to dispatch a method invocation to <i>thePProcKType</i>, which is an object in a dynamically loaded module. To guard the dispatch, the virtual call is performed inside a <i>try</i> block.</p>

<p>The system call catches all <i>ProntoExceptions</i> and calls the virtual function report. The packet processors can throw subclasses of <i>ProntoException</i> and customize the <i>report</i> function. Finally, all other exceptions are caught with the second clause. This could include processor-level errors if the OS provides support for mapping processor-level errors into <i>catch</i>able exceptions.</p>
<h3>Using Try Blocks in the Data Path</h3>

<p>The Pronto data path consists of a classifier that maps packets to flows. Each flow is associated with a forwarding path consisting of chains of packet processors. Each forwarding path may have multiple branches. Examples of packet processors include basic IP forwarding, tunnel entry/exit, NAT functionality, and more. Packet processors are dynamically added to the router at runtime. </p>

<p><a name="rl5"></a><a href="#l5">Listing Five</a> shows how we employ exceptions in this critical part of the data path. A <i>try</i> block guards the processing of a packet as it is sent through the chain of packet processors associated with the flow they belong to (identified by the call to the classifier above the <i>try</i>). As in the previous example, there are two catch statements, one catching all <i>ProntoExceptions</i>, the other catching all.</p>

<p>It is worth noting, in this example, that as new types of packet processors, say for example IPSec tunnel entry, are introduced, they may in turn introduce new subclasses of the <i>ProntoException</i>, defining a new handler (the report method). This way, the Pronto data path is capable of performing type-specific exception handling for new, dynamically installed types. </p>
<h3>Evaluation</h3>

<p>For the purpose of measuring the absolute cost of throwing an exception, we implemented a kernel module that throws an integer out of a function. Our measurements were performed on an Intel Pentium 3, 996.859 MHz running the Linux 2.6.6 kernel that has been patched to include Pronto and the C++ runtime library.</p>

<p>To put the numbers into context, we measured the performance of the Linux <i>printk</i> function, which is commonly used in exceptional circumstances to communicate error messages. The time to print a string of length 6&#151;<i>printk("Error\n")</i>&#151;was measured to be 18.15 ms. Typical usage of <i>printk</i> include formatting the strings, which is even more expensive.</p>

<p>As expected, the cost of exceptions is dependent on the number of stack frames that the exception is thrown through. <a name="rt1"></a><a href="0509ft1.html">Table 1</a> tabulates how the number of stack frames affects the cost of throwing an exception, using our optimized implementation.</p>

<p>We observe that cost increases about 0.35 ms with each stack frame. However,  when using other types of error-handling techniques, the cost also increases with the number of stack frames traversed. </p>

<p>For comparison, the cost of using exceptions without our optimizations is tabulated in <a name="rt2"></a><a href="0509ft2.html">Table 2</a>. We observe that the increase in cost for each stack frame in the GNU g++ implementation without our optimizations is around 2.5 ms. Hence, the effect of our optimization is even more impressive when throwing the exception through multiple functions. </p>

<p>Finally, we measured the effect of using <i>try</i> blocks in the data path when no exceptions occur. We observed an increase of 2.8 percent in packet latency&#151;consistent with the results of other writers. Using exceptions induces a slight overhead, even though no instructions are executed on entry to <i>try</i> blocks, which can be accredited to less successful optimizations of compilers.</p>
<h3>Summary</h3>

<p>In this article, we discussed our C++ kernel-level runtime support for Linux, which lets you use the full power of C++ in kernel-space programming, including global constructors and destructors, dynamic type checking, and exceptions.</p>


<p><b>DDJ</b></p>



	
	
<br>
<br>
<b><a name="l1"></a>Listing One</b><br>
<pre>struct task_struct {
   volatile long state;
   struct thread_info *thread_info;
   ...
#ifdef CONFIG_CXX_RUNTIME
   struct {
     void *caughtExceptions;
     unsigned int uncaughtExceptions;
   } cxa_eh_globals;
#endif
}; </pre>
<a href="#rl1">Back to article</a><br>
	
	
	
<br>
<br>
<b><a name="l2"></a>Listing Two</b><br>
<pre>#include &lt;begin_include.h&gt;
#include &lt;linux/module.h&gt;
#include &lt;linux/kernel.h&gt;
#include &lt;end_include.h&gt;</pre>
<a href="#rl2">Back to article</a><br>
	
	
	
<br>
<br>
<b><a name="l3"></a>Listing Three</b><br>
<pre>class OSException 
{ 
public: 
     char* getMessage();      
     OSException(char* msg,int sev);
     int getSeverity(); 
     virtual void report(); 
     enum tSeverity {MINOR=1,MAJOR,FATAL}; 
private: 
     char* message; 
     int severity; 
}; 
class NetworkException : public OSException 
{ 
   ...
}; 
class ProntoException : public NetworkException 
{
   ...
}; </pre>
<a href="#rl3">Back to article</a><br>
	
	
	
<br>
<br>
<b><a name="l4"></a>Listing Four</b><br>
<pre>asmlinkage int 
sys_pproc_type_call(int pptype, int call, void* args) 
{ 
  int retval = -ENOSYS; 
  try { 
    if (thePProcKType) {
      retval = 
         thePProcKType-&gt;syscall(pptype, call, args); 
    } else { 
      printk(KERN_ERR "pproc not loaded");   
    } 
  } catch(ProntoException &amp; exception) { 
         exception.report(); 
  } catch(...) { 
        printk(KERN_ERR "Unknown Exception occurred"); 
  } 
  return retval; 
} </pre>
<a href="#rl4">Back to article</a><br>
	
	
	
<br>
<br>
<b><a name="l5"></a>Listing Five</b><br>
<pre>int pronto_ip_rcv(struct sk_buff *skb, .) 
{ 
  ...  
  flow = classifier-&gt;lookup(skb); 
  if( flow ){
    try { 
       flow-&gt;arrive( skb ); 
    } catch(ProntoException &amp; exception) { 
       exception.report(); 
       kfree_skb(skb); 
    } catch(...) { 
     printk("Unknown exception occurred"); 
     kfree_skb(skb); 
    } 
  }
  ...
} </pre>
<a href="#rl5">Back to article</a><br>
	
	
	


</body>
</html>